yarn-project/noir-libs/value-note/src/utils.nr

```diff
@@ -1,184 +1,125 @@
+use dep::std::option::Option;
 use dep::aztec::context::PrivateContext;
 // docs:start:encrypted_import
 
 use dep::aztec::log::emit_encrypted_log;
 
 // docs:end:encrypted_import
-use dep::aztec::note::note_getter_options::NoteGetterOptions;
+use dep::aztec::note::note_getter_options::{NoteGetterOptions, SortOrder};
 use dep::aztec::oracle::get_public_key::get_public_key;
 use dep::aztec::state_vars::set::Set;
-use dep::aztec::types::point::Point;
 use crate::{
     filter::filter_notes_min_sum,
     value_note::{ValueNote, VALUE_NOTE_LEN},
 };
 
-fn spend_notes(
+// Sort the note values (0th field) in descending order.
+// Pick the fewest notes whose sum is equal to or greater than `amount`.
+fn create_note_getter_options_for_decreasing_balance(amount: Field) -> NoteGetterOptions<ValueNote, VALUE_NOTE_LEN, Field> {
+    NoteGetterOptions::with_filter(filter_notes_min_sum, amount).sort(0, SortOrder.DESC)
+}
+
+// Creates a new note for the recipient.
+// Inserts it to the recipient's balance.
+fn increase_balance(
     context: &mut PrivateContext,
     balance: Set<ValueNote, VALUE_NOTE_LEN>,
     amount: Field,
-    owner: Field,
+    recipient: Field,
 ) {
-    let options = NoteGetterOptions::with_filter(filter_notes_min_sum, amount);
-    let opt_notes = balance.get_notes(context, options);
-
-    let mut sum = 0;
-    for i in 0..opt_notes.len() {
-        if opt_notes[i].is_some() {
-            let note = opt_notes[i].unwrap_unchecked();
-
-            // Ensure the notes are actually owned by the owner (to prevent user from generating a valid proof while
-            // spending someone else's notes).
-            assert(note.owner == owner);
-
-            // Removes the note from the owner's set of notes.
-            balance.remove(context, note);
-
-            sum += note.value;
-        }
-    }
-
-    assert(sum as u120 >= amount as u120);
-
-    // Creates change note for the owner.
-    let change_value = sum - amount;
-    let mut change_note = ValueNote::new(change_value, owner);
-    balance.insert(context, &mut change_note);
-
-    // Emit the newly created encrypted note preimages via oracle calls.
-    let mut encrypted_data = [0; VALUE_NOTE_LEN];
-    if change_value != 0 {
-        encrypted_data = change_note.serialise();
-    };
+    let mut note = ValueNote::new(amount, recipient);
+    create_note(context, balance, recipient, &mut note);
 
-    let encryption_pub_key = get_public_key(owner);
-    emit_encrypted_log(
-        context,
-        (*context).this_address(),
-        balance.storage_slot,
-        encryption_pub_key,
-        encrypted_data,
-    );
+    // It won't compile if Set.insert() is in an if statement :(
+    // if amount as u120 > 0 {
+    //     create_note(context, balance, recipient, &mut note);
+    // }
```

Comment:
Might it be possible to come up with a minimal example which demonstrates this issue, to provide to the Noir team?

Reply:
I tried it for quite a while but couldn't. Will try again when I get my current PRs merged!